### PR TITLE
DAOS-16329 chk: maintenance mode after checking pool with dryrun

### PR DESCRIPTION
**Pull request by Nasf-Fan**
Sometimes, after system shutdown unexpectedly, the users may expect to check their critical data under some kind of maintenance mode. Under such mode, no user data can be modified or moved or aggregated. That will guarantee no further potential (DAOS logic caused) damage can happen during the check.

For such purpose, we will enhance current DAOS CR logic with --dryrun option to allow the pool (after check) to be opened as immutable with disabling some mechanism that may potentially cause data modification or movement (such as rebuild or aggregation).

Under such mode, if client wants to connect to the pool, the read-only option must be specified. Similarly for opening container in such pool.

Test-tag: pr cat_recov
Allow-unstable-test: true

Signed-off-by: Fan Yong <fan.yong@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/cart/util/cart_logtest.py
Comment (daltonbohning) [inline src/tests/ftest/cart/util/cart_logtest.py:212]: Why are we ignoring `cont_open` here instead of fixing the errors? And what are the errors we're ignoring?

**Review by daltonbohning — DISMISSED** [review]

File: src/security/srv_acl.c
Comment (kjacque) [inline src/security/srv_acl.c:503]: simpler than this, I wonder about checking this at the handler level:
```c
if (immutable && (flags != DAOS_PC_RO)) {
        D_ERROR("caller requested RW permissions on immutable pool\n");
        return -DER_NO_PERM;
}
```
Then I don't think any change would be needed in the security capas code at all. If you make the simpler check early in the RPC handlers, only RO requests will get through to this point, so the capas won't need to be further filtered.

**Review by kjacque — CHANGES_REQUESTED** [review]

File: src/chk/chk_engine.c
Comment (Nasf-Fan) [reply src/chk/chk_engine.c:1761]: You are right, seems that we have to evict the old handles. I will fix it.

File: src/security/srv_acl.c
Comment (Nasf-Fan) [reply src/security/srv_acl.c:503]: Yes, I have ever implemented like that, it can work. But as former discussion, we hoped to unify the check `if (immutable && (flags != DAOS_PC_RO))` into `ds_sec_pool_can_connect()`. That will change `ds_sec_pool_can_connect()`. So either `ds_sec_pool_get_capabilities()` or `ds_sec_pool_can_connect()` needs some enhancement to support `immutable` feature. I prefer to change `ds_sec_pool_get_capabilities()`. Because if still want to implement my original idea about the security capability `ds_sec_pool_get_capabilities()` in future, then we do not need to re-change related security interfaces.

Comment (kjacque) [reply src/security/srv_acl.c:503]: On seeing the code, I guess the question for me is whether it is useful to unify the immutable check into the capas calculation at all. We don't even need to get the ACL/owner props to know the user has made an invalid request, asking for RW permissions to an immutable pool. It feels almost like an input consistency check, rather than a capability calculation.

Comment (Nasf-Fan) [reply src/security/srv_acl.c:503]: OK, I will move `immutable` check out side of security module, instead, related RPC handler (for pool connection and container open) will handle that explicitly.

**Review by wangshilong — DISMISSED** [review]

**Review by kjacque — APPROVED** [review]

**Review by wangshilong — APPROVED** [review]

**Review by daltonbohning — COMMENTED** [review]
Will reserve approval for those familiar with core code, I think we will need to force land for the existing NLT failure

**Comment by daltonbohning** [comment]
I'm not familiar enough to merge myself without reviews from relevant @daos-stack/metadata-owners and @daos-stack/client-api-owners. Will leave this to another gatekeeper or let those owners approve first.

**Comment by Nasf-Fan** [comment]
> I'm not familiar enough to merge myself without reviews from relevant @daos-stack/metadata-owners and @daos-stack/client-api-owners. Will leave this to another gatekeeper or let those owners approve first.

Ping @gnailzenh @NiuYawei @daos-stack/daos-gatekeeper , thanks!

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**
